src/qibo/backends/tensorflow.py

```diff
@@ -0,0 +1,150 @@
+import os
+from qibo.backends import numpy
+from qibo.config import LOG_LEVEL, raise_error
+
+
+class Optimization:
+
+    def __init__(self):
+        import tensorflow as tf
+        self.Variable = tf.Variable
+        self.GradientTape = tf.GradientTape
+        self.optimizers = tf.optimizers
+
+
+class TensorflowBackend(numpy.NumpyBackend):
+
+    def __init__(self):
+        super().__init__()
+        os.environ["TF_CPP_MIN_LOG_LEVEL"] = str(LOG_LEVEL)
+        import tensorflow as tf
+        self.backend = tf
+        self.name = "tensorflow"
+
+        self.cpu_devices = tf.config.list_logical_devices("CPU")
+        self.gpu_devices = tf.config.list_logical_devices("GPU")
+        if self.gpu_devices: # pragma: no cover
+            # CI does not use GPUs
+            self.default_device = self.gpu_devices[0].name
+        elif self.cpu_devices:
+            self.default_device = self.cpu_devices[0].name
+        else: # pragma: no cover
+            # case not tested by GitHub workflows because it requires no device
+            raise_error(RuntimeError, "Unable to find Tensorflow devices.")
+
+        from qibo.backends import matrices
+        self.matrices = matrices.TensorflowMatrices(self.dtypes('DTYPECPX'))
+
+        self.tensor_types = (self.np.ndarray, tf.Tensor, tf.Variable)
+        self.Tensor = tf.Tensor
+        self.random = tf.random
+        self.newaxis = tf.newaxis
+        from tensorflow.python.framework import errors_impl # pylint: disable=E0611
+        self.oom_error = errors_impl.ResourceExhaustedError
+        self.optimization = Optimization()
+
+        from qibo.tensorflow import custom_operators as op
```

Comment:
If I understand correctly, `custom/defaulteinsum/matmuleinsum` require custom operators, right? I see that the initial_state operator is always used.

Reply:
Yes, that's right, all Tensorflow backends use the same custom operator for the initial state. This is the case in our current master too. In principle we could write the state initialization using Tensorflow primitives for the defaulteinsum/matmul backend. This way it would be possible to use these backends without compiling the custom operators. I don't think there will be much change in performance since these backends are much slower than custom anyway. If you agree we can implement this, it is relatively simple to do.

The numpy backends already have their own initial state:
```Python
state = np.zeros(2 ** nqubits)
state[0] = 1
```
so that they are completely custom operator free, in order to be usable in an environment without Tensorflow.

Reply:
Ok, we can keep for the time being the custom init for all tf backends.